runtime/ts/plan/planning-result.ts

```diff
@@ -35,6 +35,69 @@ export class PlanningResult {
     this._suggestions = suggestions;
   }
 
+  static formatSerializableGenerations(generations) {
+    // Make a copy of everything and assign IDs to recipes.
+    const idMap = new Map(); // Recipe -> ID
+    let lastID = 0;
+    const assignIdAndCopy = recipe => {
+      idMap.set(recipe, lastID);
+      const {result, score, derivation, description, hash, valid, active, irrelevant} = recipe;
+      const resultString = result.toString({showUnresolved: true, showInvalid: false, details: ''});
+      const isResolved = result.isResolved();
+      return {resultString, isResolved, score, derivation, description, hash, valid, active, irrelevant, id: lastID++};
+    };
+    generations = generations.map(pop => ({
+      record: pop.record,
+      generated: pop.generated.map(assignIdAndCopy)
+    }));
+
+    // Change recipes in derivation to IDs and compute resolved stats.
+    return generations.map(pop => {
+      const population = pop.generated;
+      const record = pop.record;
+      // Adding those here to reuse recipe resolution computation.
+      record.resolvedDerivations = 0;
+      record.resolvedDerivationsByStrategy = {};
+
+      population.forEach(item => {
+        item.derivation = item.derivation.map(derivItem => {
+          let parent;
+          let strategy;
+          if (derivItem.parent) {
+            parent = idMap.get(derivItem.parent);
+          }
+          if (derivItem.strategy) {
+            strategy = derivItem.strategy.constructor.name;
+          }
+          return {parent, strategy};
+        });
+        item.resolved = item.isResolved;
+        if (item.resolved) {
+          record.resolvedDerivations++;
+          const strategy = item.derivation[0].strategy;
+          if (record.resolvedDerivationsByStrategy[strategy] === undefined) {
+            record.resolvedDerivationsByStrategy[strategy] = 0;
+          }
+          record.resolvedDerivationsByStrategy[strategy]++;
+        }
+        const options = {showUnresolved: true, showInvalid: false, details: ''};
+        item.result = item.resultString;
```

Comment:
Why have both `item.result` and `item.resultStirng`? Maybe revert to what was previously here (i.e. calculate result in place here) or write to `item.result` directly in `assignIdAndCopy`, instead of `item.resultString`.

Reply:
good point, removed and cleanup a bit. thanks!